src/som/interpreter/nodes/dispatch/BlockDispatchNode.java

```diff
@@ -0,0 +1,64 @@
+package som.interpreter.nodes.dispatch;
+
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.DirectCallNode;
+import com.oracle.truffle.api.nodes.IndirectCallNode;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.object.DynamicObject;
+
+import som.interpreter.SArguments;
+import som.vmobjects.SBlock;
+import som.vmobjects.SInvokable;
+import som.vmobjects.SInvokable.SMethod;
+
+
+public abstract class BlockDispatchNode extends Node {
+
+  public abstract Object executeDispatch(VirtualFrame frame, Object[] arguments);
+
+  protected static final boolean isSameMethod(final Object[] arguments,
+      final DynamicObject cached) {
+    if (!(arguments[0] instanceof SBlock)) {
+      return false;
+    }
+    return getMethod(arguments) == cached;
+  }
+
+  protected static final DynamicObject getMethod(final Object[] arguments) {
+    DynamicObject method = ((SBlock) arguments[0]).getMethod();
+    assert SMethod.getNumberOfArguments(method) == arguments.length;
+    return method;
+  }
+
+  protected static final DirectCallNode createCallNode(final Object[] arguments,
+      final VirtualFrame frame) {
+    return Truffle.getRuntime().createDirectCallNode(
+        SInvokable.getCallTarget(getMethod(arguments), SArguments.getExecutionLevel(frame)));
+  }
+
+  @Specialization(guards = "isSameMethod(arguments, cached)")
+  public Object activateBlock(final VirtualFrame frame, final Object[] arguments,
+      @Cached("getMethod(arguments)") final DynamicObject cached,
+      @Cached("createCallNode(arguments, frame)") final DirectCallNode call) {
+    return call.call(SArguments.createSArguments(SArguments.getEnvironment(frame),
```

Comment:
What about intercession points?
No MOP support for any of this?

Reply:
These BlockDispatchs are inside a primitive. in Mate, at specialization time, this primitives are wrapped with a MateEager*PrimitiveNode that installs an IH for message lookup and message activation.